File: tools/wasm/src_js/sync/query_result.js
/**
 * @file QueryResult stores the result of a query execution.
 */
"use strict";

const KuzuWasm = require("./kuzu.js");

class QueryResult {
  /**
   * Internal constructor. Use `Connection.query` or `Connection.execute`
   * to get a `QueryResult` object.
   * @param {kuzu.sync.QueryResult} _queryResult the native query result object.
   */
  constructor(_queryResult, _isClosable = true) {
    this._result = _queryResult;
    this._isClosed = false;
    this._isClosable = _isClosable;
  }

  /**
   * Internal function to check if the query result is closed.
   * @throws {Error} if the query result is closed.
   * @private
   */
  _checkQueryResult() {
    KuzuWasm.checkInit();
    if (this._isClosed) {
      throw new Error("QueryResult is already closed.");
    }
  }

  /**
   * Check if the query result is successfully executed.
   * @returns {Boolean} true if the query result is successfully executed.
   * @throws {Error} if the query result is closed.
   */
  isSuccess() {
    this._checkQueryResult();
    return this._result.isSuccess();
  }

  /**
   * Get the error message if the query result is not successfully executed.
   * @returns {String} the error message.
   * @throws {Error} if the query result is closed.
   */
  getErrorMessage() {
    this._checkQueryResult();
    return this._result.getErrorMessage();
  }

  /**
   * Reset the iterator of the query result to the beginning.
   * This function is useful if the query result is iterated multiple times.
   * @throws {Error} if the query result is closed.
   */
  resetIterator() {
    this._checkQueryResult();
    this._result.resetIterator();
  }

  /**
   * Get the number of rows of the query result.
   * @returns {Number} the number of rows of the query result.
   * @throws {Error} if the query result is closed.
   */
  hasNext() {
    this._checkQueryResult();
    return this._result.hasNext();
  }

  /**
   * Check if the query result has a following query result when multiple 
   * statements are executed within a single query.
   * @returns {Boolean} true if the query result has a following query result.
   * @throws {Error} if the query result is closed.
   */
  hasNextQueryResult() {
    this._checkQueryResult();
    return this._result.hasNextQueryResult();
  }

  /**
   * Get the number of columns of the query result.
   * @returns {Number} the number of columns of the query result.
   * @throws {Error} if the query result is closed.
   */
  getNumColumns() {
    this._checkQueryResult();
    return this._result.getNumColumns();
  }

  /**
   * Get the number of rows of the query result.
   * @returns {Number} the number of rows of the query result.
   * @throws {Error} if the query result is closed.
   */
  getNumTuples() {
    this._checkQueryResult();
    return this._result.getNumTuples();
  }

  /**
   * Get the column names of the query result.
   * @returns {Array<String>} the column names of the query result.
   * @throws {Error} if the query result is closed.
   */
  getColumnNames() {
    this._checkQueryResult();
    return this._result.getColumnNames();
  }

  /**
   * Get the column types of the query result.
   * @returns {Array<String>} the column types of the query result.
   * @throws {Error} if the query result is closed.
   */
  getColumnTypes() {
    this._checkQueryResult();
    return this._result.getColumnTypes();
  }

  /**
   * Get the string representation of the query result.
   * @returns {String} the string representation of the query result.
   * @throws {Error} if the query result is closed.
   */
  toString() {
    this._checkQueryResult();
    return this._result.toString();
  }

  /**
   * Get the query summary (execution time and compiling time) of the query 
   * result.
   * @returns {Object} the query summary of the query result.
   * @throws {Error} if the query result is closed.
   */
  getQuerySummary() {
    this._checkQueryResult();
    return this._result.getQuerySummary();
  }

  /**
   * Get the following query result when multiple statements are executed within 
   * a single query.
   * @returns {QueryResult} the next query result.
   * @throws {Error} if the query result is closed.
   */
  getNextQueryResult() {
    this._checkQueryResult();
    return new QueryResult(this._result.getNextQueryResult(), false);
  }

  /**
   * Get the next row of the query result.
   * @returns {Array} the next row of the query result.
   * @throws {Error} if the query result is closed.
   */
  getNext() {
    this._checkQueryResult();
    return this._result.getNext();
  }

  /**
   * Get all rows of the query result.
   * @returns {Array<Array>} all rows of the query result.
   * @throws {Error} if the query result
   */
  getAllRows() {
    this._checkQueryResult();
    return this._result.getAsJsNestedArray();
  }

  /**
   * Get all objects of the query result.
   * @returns {Array<Object>} all objects of the query result.
   * @throws {Error} if the query result is closed.
   */
  getAllObjects() {
    this._checkQueryResult();
    return this._result.getAsJsArrayOfObjects();
  }

  /**
   * Close the query result.
   */
  close() {
    if (!this._isClosed) {
      if (this._isClosable) {
        this._result.delete();
      }
      this._isClosed = true;
    }
  }
}

module.exports = QueryResult;

File: tools/java_api/src/main/java/com/kuzudb/QueryResult.java
package com.kuzudb;

/**
 * QueryResult stores the result of a query execution.
 */
public class QueryResult implements AutoCloseable {
    long qr_ref;
    boolean destroyed = false;
    boolean isOwnedByCPP = false;

    /**
     * Check if the query result has been destroyed.
     *
     * @throws RuntimeException If the query result has been destroyed.
     */
    private void checkNotDestroyed() {
        if (destroyed)
            throw new RuntimeException("QueryResult has been destroyed.");
    }

    /**
     * Close the query result and release the underlying resources. This method is invoked automatically on objects managed by the try-with-resources statement.
     *
     * @throws RuntimeException If the query result has been destroyed.
     */
    @Override
    public void close() {
        destroy();
    }

    public boolean isOwnedByCPP() {
        return isOwnedByCPP;
    }

    /**
     * Destroy the query result.
     *
     * @throws RuntimeException If the query result has been destroyed.
     */
    private void destroy() {
        checkNotDestroyed();
        if (!isOwnedByCPP) {
            Native.kuzuQueryResultDestroy(this);
            destroyed = true;
        }
    }

    /**
     * Check if the query is executed successfully.
     *
     * @return Query is executed successfully or not.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public boolean isSuccess() {
        checkNotDestroyed();
        return Native.kuzuQueryResultIsSuccess(this);
    }

    /**
     * Get the error message if any.
     *
     * @return Error message of the query execution if the query fails.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public String getErrorMessage() {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetErrorMessage(this);
    }

    /**
     * Get the number of columns in the query result.
     *
     * @return The number of columns in the query result.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public long getNumColumns() {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetNumColumns(this);
    }

    /**
     * Get the column name at the given index.
     *
     * @param index: The index of the column.
     * @return The column name at the given index.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public String getColumnName(long index) {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetColumnName(this, index);
    }

    /**
     * Get the column data type at the given index.
     *
     * @param index: The index of the column.
     * @return The column data type at the given index.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public DataType getColumnDataType(long index) {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetColumnDataType(this, index);
    }

    /**
     * Get the number of tuples in the query result.
     *
     * @return The number of tuples in the query result.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public long getNumTuples() {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetNumTuples(this);
    }

    /**
     * Get the query summary.
     *
     * @return The query summary.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public QuerySummary getQuerySummary() {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetQuerySummary(this);
    }

    /**
     * Return if the query result has next tuple or not.
     *
     * @return Whether there are more tuples to read.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public boolean hasNext() {
        checkNotDestroyed();
        return Native.kuzuQueryResultHasNext(this);
    }

    /**
     * Get the next tuple. Note that to reduce resource allocation, all calls to
     * getNext() reuse the same FlatTuple object. Since its contents will be
     * overwritten, please complete processing a FlatTuple or make a copy of its
     * data before calling getNext() again.
     *
     * @return The next tuple.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public FlatTuple getNext() {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetNext(this);
    }

    /**
     * Return if the query result has next query result or not.
     *
     * @return Whether there are more query results to read.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public boolean hasNextQueryResult() {
        checkNotDestroyed();
        return Native.kuzuQueryResultHasNextQueryResult(this);
    }

    /**
     * Get the next query result.
     *
     * @return The next query result.
     * @throws RuntimeException If the query result has been destroyed.
     */
    public QueryResult getNextQueryResult() {
        checkNotDestroyed();
        return Native.kuzuQueryResultGetNextQueryResult(this);
    }

    /**
     * Convert the query result to string.
     *
     * @return The string representation of the query result.
     */
    public String toString() {
        if (destroyed)
            return "QueryResult has been destroyed.";
        else
            return Native.kuzuQueryResultToString(this);
    }

    /**
     * Reset the query result iterator.
     *
     * @throws RuntimeException If the query result has been destroyed.
     */
    public void resetIterator() {
        checkNotDestroyed();
        Native.kuzuQueryResultResetIterator(this);
    }
}
